### ISSUE TITLE
Segmentation fault with newer versions of zlib (1.1.12)

### ISSUE THREAD

**Issue by EricR86** [PyTables/PyTables#686]
This issue was initially described here:
https://github.com/conda-forge/pytables-feedstock/issues/29

On systems with newer versions of zlib installed the following will segfault:
```python
from tables import utilsextension
import zlib
zlib.decompress(zlib.compress(b'foo'))
```

This issue is reproducible on any environment that has later versions of zlib installed. I can produce this with Python 2 and 3. If you remove the tables import, or re-arrange it so it comes after, the segfault will not happen.

This issue is related to:
https://github.com/conda-forge/pytables-feedstock/issues/27
https://github.com/PyTables/PyTables/issues/644
https://github.com/PyTables/PyTables/issues/662
https://github.com/ContinuumIO/anaconda-issues/issues/7315

Here's why this is happening:
PyTables has a builtin version of zlib (1.2.8) compiled into it's shared object libraries. In the example I show in the original issue, the shared libraries are loaded in this order:

`utilsextensions.so` -- dynamically loads --> `libhdf5.so` -- dynmaically loads --> `libz.so`

When Python reconciles which functions (or symbols) to use from libz.so, it sees that `utilsextensions.so` already has them, and is earlier in the dependency chain. So Python uses `utilsextensions.so` own compiled-in version of zlib (1.2.8). When it comes to Python using the standard library "zlib", it dynamically loads it's own `zlib.so` which dynamically loads `libz.so`. The symbols at this point already resolve to the ones in `utilsextension.so` so they are used instead. This causes a segfault.

To see the details of how I came to this result you can look at the initial issue posted above.

A few solutions proposed:
1. Have the shared objects built by PyTables dynmically load in zlib instead, instead of compiling/vendoring in a specific version. There's a permanent risk here currently where there's a hope that the builtin version remains binary compatible with the one in the current running environment.

2. If versioning is absolutely needed, have PyTables create it's own shared object version of zlib (versioned correctly). The RPATH of `utilsextension.so` et al. can even point explicitly to this zlib shared object. This way there's no confusion when it comes to the dynamic loader (at least on linux). It can (hopefully) reconcile what version each shared object needs and there's no odd symbol resolution to older or newer versions.

I haven't explictly tested but since newer versions of HDF5 require newer versions of zlib, it's possible that calls into the HDF5 shared library will cause segfaults if you knew which functions to call since it attempts to dynamically load and use the system `libz.so` and not the builtin version provided by PyTables. I don't believe this issue can be simply "fixed" by users re-arranging their import order since there will be a risk of running into an issue regardless.

**Comment by avalentino** [comment]
Strictly speaking the only embedded library in the pytables source distribution is c-blosc that brings with it a set of compression libraries (located in c-blosc/internal-complibs/).

The PyTables setup.py script uses the embedded copy of the c-blosc as a fallback if a reasonably updated version of the blosc library is not found in the system.

IMO in a system like anaconda PyTables shall be built against the anaconda version of blosc and all dependencies and compatibility between libraries shall be ensured by the package manager.

In this way there will be a single version of e.g. zlib in the anaconda environment an all packages, including pytables, will use it.

In short IMO the problem could be fixed by adding blosc to the requirements (build and run) of the pytables-feedstock recipe.

**Comment by avalentino** [comment]
I'm closing this issue because it seems to be a packaging issue more than a PyTables bug (see also #644).
Anyway PR for

* additional recommendations for packagers in the installation guide (PyTables/doc/source/usersguide/installation.rst)
* some machinery to allow to force the use system libraries in the setup.py script

would be very welcome
